lib/sigstore/signer.rb

```diff
@@ -0,0 +1,246 @@
+      # compatibility with some signing metadata formats (see §Submission of Signing Metadata to Transparency Service).
+      key.sign("SHA256", payload)
+    end
+
+    # TODO: implement
+    def submit_signature_hash_to_timstamping_service(_signature)
+      # The Signer sends a hash of the signature as the messageImprint in a TimeStampReq to the Timestamping Service and
+      # receives a TimeStampResp including a `TimeStampToken`.
+      # The signer MUST verify the TimeStampToken against the payload and Timestamping Service root certificate.
+
+      {}
+    end
+
+    def submit_signed_metadata_to_transparency_service(signature, cert, data_sha256)
+      # The Signer chooses a format for signing metadata; this format MUST be in the supportedMetadataFormats in the
+      # Transparency Service configuration. The Signer prepares signing metadata containing at a minimum:
+      # * The signature.
+      # * The payload (possibly pre-hashed; if so, the entry also includes the identifier of the hash algorithm).
+      # * Verification material (signing certificate or verification key).
+      #   * If the verification material is a certificate, the client SHOULD upload only the signing certificate and
+      #     SHOULD NOT upload the CA certificate chain.
+      #
+      # The signing metadata might contain additional, application-specific metadata according to the format used.
+      # The Signer then canonically encodes the metadata (according to the chosen format).
+
+      # TODO: is looping here correct?
+      @trusted_root.tlogs_for_signing.map do |ctlog|
+        logger.info { "Submitting to #{ctlog.base_url}" }
+
+        body = {
+          "spec" => {
+            "signature" => {
+              "content" => Internal::Util.base64_encode(signature),
+              "publicKey" => {
+                "content" => Internal::Util.base64_encode(cert.to_pem)
+              }
+            },
+            "data" => {
+              "hash" => {
+                "algorithm" => "sha256", # TODO: should this always be sha256?
+                "value" => Internal::Util.hex_encode(data_sha256)
+              }
+            }
+          },
+          "kind" => "hashedrekord", # TODO: is hashedrekord always the right kind? should this be configurable?
+          "apiVersion" => "0.0.1"
+        }
+
+        resp = Net::HTTP.post(URI.join(ctlog.base_url, "api/v1/log/entries"), body.to_json, {
+                                "Content-Type" => "application/json"
+                              })
+
+        logger.debug do
+          "#{resp.code} #{resp.message.inspect}\n\n#{JSON.pretty_generate(body)}\n\n#{resp.body}"
+        end
+
+        unless resp.code == "201"
+          raise Error::Signing, "#{resp.code} #{resp.message.inspect}\n\n#{JSON.pretty_generate(body)}\n\n#{resp.body}"
+        end
+
+        body = JSON.parse(resp.body)
+
+        # TODO: verify
```

Comment:
- [ ] do we need to verify the returned log entry separately from the full verification flow?

Reply:
I don't think so; it should be embedded just into the bundle.